### PR TITLE
Introduce more usages to describe things related with the states of LRO

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Remove this workaround since #311 is closed.
Fixes https://github.com/Azure/typespec-azure/issues/1752
Fixes https://github.com/Azure/typespec-azure/issues/1530

This PR adds three new usage flags: `LroInitial`, `LroPolling`, `LroFinalEnvelope`.
These new usage flags are previously represented by `Output`, now the `Output` flag no longer contain the semantics of these usages.
Therefore, if a model's usage is only `LroInitial`, this means it is only used as the initial response of an LRO.
if a model has both `LroInitial` and `Output`, this means it is used as the initial response of an LRO, and it also used as a normal output in other places.
The final result will not have above new usages unless it participates as those roles as well. It will have output usage.

### CODE REVIEW

File: packages/typespec-client-generator-core/test/methods/lro.test.ts [hunk 1]
```diff
@@ -359,18 +369,48 @@ describe("typespec-client-generator-core: long running operation metadata", () =
...
         ok(pollingModel);
+        strictEqual(pollingModel.usage & UsageFlags.Input, 0); // polling model should not be input
+        notStrictEqual(pollingModel.usage & UsageFlags.Output, 0); // polling model should be output
         strictEqual(metadata.pollingStep.responseBody, pollingModel);
 
         assert.isUndefined(metadata.finalResponse);
       });
     });
+
+    it("LRO defined in different namespace", async () => {
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/test/methods/lro.test.ts:387]: for this [issue](https://github.com/Azure/typespec-azure/issues/1752), two namespaces do not have hierarchy. could you change your test to that scenario.

**Review by tadelesh — APPROVED** [review]

File: packages/typespec-client-generator-core/test/methods/lro.test.ts (hunk 1 above)
Comment (ArcturusZhang) [reply packages/typespec-client-generator-core/test/methods/lro.test.ts:387]: updated

**Comment by iscai-msft** [comment]
@ArcturusZhang are you ok with this being released as part of the november release? in 2 weeks? If so, I would merge into main. If you'd like us to release this before the big release, can you merge into our release branch `release/november-2024` and let us know, then we'll do a tcgc release for you. Thanks for your contribution!

**Comment by haolingdong-msft** [comment]
The changes in this pr looks good to me. Just want to bring up this related issue: https://github.com/Azure/typespec-azure/issues/1750 - If TCGC keeps the current usage and access for LRO intial response, we should have a way to let emitter know how to update access and do not generate this model.

**Comment by ArcturusZhang** [comment]
> The changes in this pr looks good to me. Just want to bring up this related issue: #1750 - If TCGC keeps the current usage and access for LRO intial response, we should have a way to let emitter know when to update access and do not generate this model. But this will make TCGC's access less helpful as emitter still need to update it based on usage.

I think per discussion this morning, maybe it is better that we could introduce some new usages for these lro related models since they have different context to use.

**Comment by haolingdong-msft** [comment]
> I think per discussion this morning, maybe it is better that we could introduce some new usages for these lro related models since they have different context to use.

Is there any issue to track? (my created issue is closed as planned)

**Comment by tadelesh** [comment]
> > I think per discussion this morning, maybe it is better that we could introduce some new usages for these lro related models since they have different context to use.
> 
> Is there any issue to track? (my created issue is closed as planned, feel free to reopen it)

i think we could do the usage split in this pr. @ArcturusZhang thoughts?

**Comment by ArcturusZhang** [comment]
> > > I think per discussion this morning, maybe it is better that we could introduce some new usages for these lro related models since they have different context to use.
> > 
> > 
> > Is there any issue to track? (my created issue is closed as planned, feel free to reopen it)
> 
> i think we could do the usage split in this pr. @ArcturusZhang thoughts?

yeah I will update those in this PR

**Review by weidongxu-microsoft — COMMENTED** [review]
I don't have opinion on this in Java. You can go ahead if no objection from other languages.

I assume the usage is additive. Meaning we cannot know whether a model is "only used in LroInitial"?

**Comment by ArcturusZhang** [comment]
> I don't have opinion on this in Java. You can go ahead if no objection from other languages.
> 
> I assume the usage is additive. Meaning we cannot know whether a model is "only used in LroInitial"?

not quite additive.
The new usage (for example `LroInitial`) is part of the old `Output`. New `Output` no longer contains `LroInitial`.
Therefore if you see one model only has `LroInitial`, this means this model is only used as the initial response of an lro. The output now means everything other than those have an explicit definitions such as the lro initial.

**Comment by weidongxu-microsoft** [comment]
> not quite additive. The new usage (for example `LroInitial`) is part of the old `Output`. New `Output` no longer contains `LroInitial`. Therefore if you see one model only has `LroInitial`, this means this model is only used as the initial response of an lro. The output now means everything other than those have an explicit definitions such as the lro initial.

Got it. So only the "final result" would still be marked as usage=Output?

nit, maybe we can call it e.g. `OutputLroInitial` (kind of grouping the "output" family)?

**Comment by ArcturusZhang** [comment]
> > not quite additive. The new usage (for example `LroInitial`) is part of the old `Output`. New `Output` no longer contains `LroInitial`. Therefore if you see one model only has `LroInitial`, this means this model is only used as the initial response of an lro. The output now means everything other than those have an explicit definitions such as the lro initial.
> 
> Got it. So only the "final result" would still be marked as usage=Output?
> 
> nit, maybe we can call it e.g. `OutputLroInitial` (kind of grouping the "output" family)?

I think it is a good idea to emphasize. what do you think on this suggestion? @tadelesh 
An issue would be there are already some usage is part of the `input` usage but not named in this way (for instance `jsonmergepatch`)

**Comment by tadelesh** [comment]
> > > not quite additive. The new usage (for example `LroInitial`) is part of the old `Output`. New `Output` no longer contains `LroInitial`. Therefore if you see one model only has `LroInitial`, this means this model is only used as the initial response of an lro. The output now means everything other than those have an explicit definitions such as the lro initial.
> > 
> > 
> > Got it. So only the "final result" would still be marked as usage=Output?
> > nit, maybe we can call it e.g. `OutputLroInitial` (kind of grouping the "output" family)?
> 
> I think it is a good idea to emphasize. what do you think on this suggestion? @tadelesh An issue would be there are already some usage is part of the `input` usage but not named in this way (for instance `jsonmergepatch`)

i'm thinking of separating the input and output usage into a second layer, which means:
1. for first layer, we label them all as output
2. for second layer, we separate output to something like:
```
enum OutputUsage {
  Normal,
  LroInitial,
  Error,
  ...
}
```
That will be a breaking change, but much extensible and we should make sure each layer orthogonal.
Loop in @iscai-msft for thoughts.

**Review by tadelesh — APPROVED** [review]

**Comment by tadelesh** [comment]
open this issue to track usage refactor: https://github.com/Azure/typespec-azure/issues/1941

**Pull request merged**